### PR TITLE
Some preparation for binary patterns

### PR DESCRIPTION
**Pull request by ice1000**
Follow up #178: improved the generalization of `BinOpParser` so it can be used in patterns more easily

### CODE REVIEW

**Comment by ice1000** [comment]
> Build succeeded:
> 
> * [gradle-check](https://github.com/aya-prover/aya-dev/runs/4201982674?check_suite_focus=true)

Thank you! Mamamia!

File: base/src/main/java/org/aya/concrete/desugar/BinExprParser.java [hunk 1]
```diff
@@ -57,7 +56,7 @@ public BinExprParser(@NotNull AyaBinOpSet opSet, @NotNull SeqView<@NotNull Elem<
     throw new IllegalArgumentException("not an operator");
   }
 
-  @Override protected @Nullable OpDecl asOpDecl(@NotNull Elem<Expr> elem) {
+  @Override protected @Nullable OpDecl asOpDecl(@NotNull Expr.NamedArg elem) {
```
Comment (imkiva) [inline base/src/main/java/org/aya/concrete/desugar/BinExprParser.java:59]: Suggest rename this method to: `referredOpDecl`

File: pretty/src/main/java/org/aya/pretty/doc/Doc.java [hunk 2]
```diff
@@ -227,6 +227,11 @@ record PageWidth(@NotNull IntFunction<Doc> docBuilder) implements Doc {
     return cat(symbol(leftSymbol), doc, symbol(rightSymbol));
   }
 
+  /** @param explicit when false, add braces */
+  static @NotNull Doc braced(Doc doc, boolean explicit) {
```
Comment (imkiva) [inline pretty/src/main/java/org/aya/pretty/doc/Doc.java:231]: Since it's in the pretty module. I suggest rename to `bracedIfNot`

**Review by imkiva — COMMENTED** [review]
Nice work!!

File: base/src/main/java/org/aya/concrete/desugar/BinExprParser.java (hunk 1 above)
Comment (ice1000) [reply base/src/main/java/org/aya/concrete/desugar/BinExprParser.java:59]: I like `underlyingOpDecl`

File: pretty/src/main/java/org/aya/pretty/doc/Doc.java (hunk 2 above)
Comment (ice1000) [reply pretty/src/main/java/org/aya/pretty/doc/Doc.java:231]: `static @NotNull Doc bracedUnless(Doc doc, boolean falsification)`

**Pull request merged**